game.py

```diff
@@ -0,0 +1,908 @@
+        
+        # return game_state
+
+    def get_next_moves(self):    
+    
+        global columns
+        global rows
+        global column_order
+                
+        moves = []
+        any_moves = False
+
+        for column in range(columns):
+            if any_moves:
+                break
+            for row in range(rows):
+                if self._board[row][column_order[column]] == None:
+                    moves.append ((row, column_order[column]))
+                    break
+                
+        return moves
+   
+    def game_end(self):
+
+# This determines whether the game has ended after the last move and whether it is a win, lose
+# or draw (if the board is full)
+
+        if self.get_number_of_moves() == 0:
+            return False, False        
+        player = - self.get_player_next()
+        board = self.get_board()
+        connect_number = self.get_connect_number()
+        last_move = self.get_last_move()
+
+        global rows
+        global columns
+        
+        game_end = False
+        win = False
+        
+        vectors = [(0, 1), (1, 0), (1, 1), (1, -1)]
+       
+        if last_move == None:
+            return game_end, win
+                     
+        for vector in vectors:
+            consec = 1
+            next_position = last_move
+            while not win:
+                next_position = (next_position[0] + vector[0], next_position[1] + vector[1])
+                # print('Next position positive : ', next_position)
+                if next_position[0] not in range(rows) or next_position[1] not in range(columns):
+                    break                                                                                    
+                if board[next_position[0]][next_position[1]] == player:
+                    consec = consec + 1
+                else:
+                    break
+                if consec == connect_number:
+                        win = True
+                # print('consec positive : ', consec, 'Vector : ', vector)
+                                        
+            next_position = last_move
+            while not win:
+                next_position = (next_position[0] - vector[0], next_position[1] - vector[1])
+                # print('Next position negative : ', next_position)
+                if next_position[0] not in range(rows) or next_position[1] not in range(columns):
+                    break                                                                                    
+                if board[next_position[0]][next_position[1]] == player:
+                    consec = consec + 1
+                else:
+                    break
+                if consec == connect_number:
+                    win = True
+                # print('consec negative : ', consec, 'Vector : ', vector)
+            
+        if win:
+            game_end = True
+ 
+        if None not in (item for sublist in self.get_board() for item in sublist):

```

Comment:
Not sure if this would speed it up (would have to dig a bit deeper) but I think this at least reads a bit better as `if any(item is None for sublist in self.get_board() for item in sublist):